### ISSUE TITLE
Problem with current Activity

### ISSUE THREAD

**Issue by renas** [RobotiumTech/robotium#495]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on March 13, 2013 04:37:24_

Hello,

I have problem with identifying current activity.
My application in test [I'm testing apk file directly] has activity A from which activity B is opened.
in setUp method I do
"solo = new Solo(getInstrumentation(), getActivity());"
and later in test

solo.clickOnText("Test1"); //Test1 text is in activity A
solo.sleep(5000); //now the application should be in activity B
solo.assertCurrentActivity("This is not correct screen!", "B"); // this fails

I tried to check activity using "solo.getCurrentActivity().toString()" and that returns "A" even if app is already in B activity

I found following issue https://code.google.com/p/robotium/issues/detail?id=179 but the workaround does not work for me
this code behaves the same

"solo = new Solo(getInstrumentation());"

at the begining of test function
getActivity();
later
solo.clickOnText("Test1"); //Test1 text is in activity A
solo.sleep(5000); //now the application should be in activity B
solo.assertCurrentActivity("This is not correct screen!", "B"); // this fails

I checked the "stack" using the "adb shell dumpsys" and from that it looks like the B is above the A including the flag "nowVisible=true"

Any idea what I'm doing wrong?

Robotium 3.6 or 4.0; Android 2.3.3 [SDK 10]
The apk file is installed and tested on real device Nexus S with Android version 4.1.2

Thanks,
  Vladimir

_Original issue: http://code.google.com/p/robotium/issues/detail?id=416_

**Comment by renas** [comment]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on March 13, 2013 04:56:31_

So it seems the problem is somehow related to the Nexus S and Android 4.1.2.
I tried the same with the Samsung Galaxy SII running Android 2.3.5 and the current activity is set correctly to B.

Is it possible that there is a problem between SDK-10 and 4.1.2?

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on March 13, 2013 06:00:32_

Thanks for reporting this. What happens if you try this on the 4.2 emulator?

**Comment by renas** [comment]
_From [miraculi...@gmail.com](https://code.google.com/u/100519236307936948879/) on March 18, 2013 08:39:45_

Same probleme here.
This problem occured on multiple devices with Android 4.0.3 or newer (didn't test Honecomb though).

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on March 18, 2013 08:42:48_

Would it be possible for you to send me your application (or another application that exhibits this issue)?

**Comment by renas** [comment]
_From [miraculi...@gmail.com](https://code.google.com/u/100519236307936948879/) on March 19, 2013 01:57:02_

I am able to reproduce a similar problem with your NotePadTest project.
I simply compile the project (Eclipse 1.5.2.20130211-1820, Java 7 x64, newest Android SDK 21.1, API17/4.2.2) and run the tests on the Samsung Galaxy Tab 2 10.1(Android 4.0.4). The test fails in assertCurrentActivity(...):


junit.framework.ComparisonFailure: Expected NoteEditor activity expected:<...Editor> but was:<...sList>
at com.jayway.android.robotium.solo.Asserter.assertCurrentActivity(Asserter.java:40)
at com.jayway.android.robotium.solo.Solo.assertCurrentActivity(Solo.java:602)
at com.jayway.test.NotePadTest.testAddNote(NotePadTest.java:47)
at java.lang.reflect.Method.invokeNative(Native Method)
at android.test.InstrumentationTestCase.runMethod(InstrumentationTestCase.java:214)
at android.test.InstrumentationTestCase.runTest(InstrumentationTestCase.java:199)
at android.test.ActivityInstrumentationTestCase2.runTest(ActivityInstrumentationTestCase2.java:192)
at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:169)
at android.test.AndroidTestRunner.runTest(AndroidTestRunner.java:154)
at android.test.InstrumentationTestRunner.onStart(InstrumentationTestRunner.java:545)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1575)

The test succeeds on the Nexus 7 (Android 4.2.2) and the Samsung Galaxy S (2.3.3), but it fails on all of my 4.0.4 devices.

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on March 19, 2013 03:15:49_

Ok. I will see if I can reproduce this. Will get back to you if I'm not able.

**Status:** Accepted  
  **Owner:** renasr...@gmail.com

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on April 07, 2013 21:15:13_

This has been corrected in Robotium 4.1.

**Status:** Verified

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on May 23, 2013 07:22:16_

Issue 462 has been merged into this issue.

**Comment by renas** [comment]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on May 23, 2013 21:17:23_

I can see the bug should be corrected in 4.1 but I met this issue with Robotium 4.1 for sure.

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on May 29, 2013 00:48:04_

Vladimir, can you explain what your issue is. Please also paste your test case.

**Comment by renas** [comment]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on May 29, 2013 00:55:23_

The problem is described in description of this bug ticket. There is no change.

Briefly
app is on Activity A
goes to another screen which is activity B
waitForActivity
assertCurrentActivity
getCurrentActivity
all of them returns activity B even after 30-60s of waiting on device with Android 4.1.2

the same code works on device with Android 2.3.5

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on May 29, 2013 00:57:09_

Do you get the same behaviour when running on 4.1.2 emulator?

**Comment by renas** [comment]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on May 29, 2013 01:20:51_

Yes, the problem is also on 4.1.2 simulator

following code failed:
// here we are in activity A
solo.sendKey(Solo.MENU);
searchText = context.getString(R.string.somemenuitem);
assertTrue(getAssertTrueText(searchText), solo.searchText(searchText));
solo.clickOnText(context.getString(R.string.somemenuitem));
assertTrue("We aren't on B activity", solo.waitForActivity(B)); //here we should be in activity B already

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on May 29, 2013 01:23:40_

Would it be possible for you to email me the application you are testing or another application that exhibits this issue?

**Comment by renas** [comment]
_From [m...@thomaskeller.biz](https://code.google.com/u/115202190558223058672/) on May 29, 2013 01:25:01_

Just for further reference - Robotium can't cope with old-school ActivityGroups and LocalActivityManagers. I had a similar problem when I tried to wrap the UI of a Google V1 MapActivity into a fragment-based application. The workaround there was to re-implement getCurrentActivity() and pick the first one on top that is not managed via a LocalActivityManager.

**Comment by renas** [comment]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on May 29, 2013 01:32:40_

@application: I could but the problem is it would not be too easy to set it up. It requires to connect to server which I don't have publicly available at the moment.

@Thomas: would it be possible that you would share your code with me? I think it might solve my issue. I guess my application uses fragments as well

**Comment by renas** [comment]
_From [renasr...@gmail.com](https://code.google.com/u/108052454046732184536/) on May 29, 2013 01:32:41_

@thomaskeller, would it be possible for you to send me the application you had the issue with?

**Comment by renas** [comment]
_From [m...@thomaskeller.biz](https://code.google.com/u/115202190558223058672/) on May 29, 2013 01:36:25_

@renasr: Unfortunately not, its closed source. The solution I found is also not a general one:

    public Activity getCurrentActivity() {
        List<Activity> openedActivities = solo.getAllOpenedActivities();
        for (int i = openedActivities.size() - 1; i >= 0; --i) {
            Intent intent = openedActivities.get(i).getIntent();
            checkArgument(intent != null, "intent must not be null");
            if (intent.getExtras() != null
                    && intent.getExtras().getBoolean(SOME_PRIVATE_MARKER)) {
                continue;
            }
            return openedActivities.get(i);
        }
        throw new IllegalStateException("could not find active activity");
    }

**Comment by renas** [comment]
_From [m...@thomaskeller.biz](https://code.google.com/u/115202190558223058672/) on May 29, 2013 01:38:11_

I forgot to add:

        Intent intent = new Intent(getActivity(), MapActivity.class);
        intent.putExtra(SOME_PRIVATE_MARKER, true);
        Window window = localActivityManager.startActivity(ACTIVITY_TAG, intent);

**Comment by renas** [comment]
_From [vladimir...@qnective.com](https://code.google.com/u/103582861998926209227/) on July 09, 2013 22:56:08_

It seems I found the root of the problem finally.
Using the same device with Android 4.1.2 one time the activity is not detected and another time it is.
The difference is how developer options are set in the phone. In case "Do not keep activities" is checked then Robotium has problem to detect activities in some cases [at least in the case described above]. If this option is not checked everything is working fine.

**Comment by renas** [comment]
_From [matt...@gmail.com](https://code.google.com/u/105453751381455581974/) on June 11, 2014 15:50:39_

I notice this issue seems to be over a year old, with no further updates? It seems to be quite a deal breaker of an issue if its not possible to progress past one Activity? Has there been a solution found? The issue from vladimir is not currently working for me.

**Comment by renas** [comment]
_From [priya.sa...@gmail.com](https://code.google.com/u/113067526213991535352/) on October 10, 2014 05:05:14_

How do I change the device time eg.samsung note 10. using robotium?

**Comment by renas** [comment]
_From [sik...@gmail.com](https://code.google.com/u/106643206084122364324/) on November 05, 2014 07:04:35_

Bump thread up!
I still get this issue in robotium 5.2.1.
I tried the work around from @thomaskeller.biz but I can get api solo.getAllOpenedActivities() anymore.

Did it get removed from 5.2.1? 
Do you guys have another solution?

Thanks!
